### PR TITLE
Protocol handler interface

### PR DESCRIPTION
**Pull request by andre4i**
# [261](https://dev.azure.com/fluidframework/internal/_workitems/edit/261/)

## Description

Part of https://github.com/microsoft/FluidFramework/pull/10412/, which showcases how this will be used. 

This change:
- extracts an interface from the protocol handler
- the interface only deals in interfaces, so we can enable any Quorum implementation
- moves a bit of logic from the container into the protocol handler so that we isolate all protocol logic into the latter.

## PR Checklist

> Use the check-list below to ensure your branch is ready for PR. If the item is not applicable, leave it blank.

-   [x] I have updated the documentation accordingly.
-   [ ] I have added tests to cover my changes.
-   [x] All new and existing tests passed.
-   [x] My code follows the code style of this project.
-   [x] I ran the lint checks which produced no new errors nor warnings for my changes.
-   [x] I have checked to ensure there aren't other open Pull Requests for the same update/change.

## Does this introduce a breaking change?

-   [ ] Yes
-   [x] No

### CODE REVIEW

**Comment by andre4i** [comment]
Looking at the scribe lambda test failures... It seems there may be a problem with the test

File: server/routerlicious/packages/protocol-base/src/protocol.ts [hunk 1]
```diff
@@ -58,16 +84,52 @@ export class ProtocolOpHandler {
...
+    snapshot(): IQuorumSnapshot {
+        return this._quorum.snapshot();
+    }
+
     public close() {
-        this.quorum.close();
+        this._quorum.close();
+    }
+
+    public processMessageInQuorum(message: ISequencedDocumentMessage, local: boolean): IProcessMessageResult {
```
Comment (vladsud) [inline server/routerlicious/packages/protocol-base/src/protocol.ts:115]: Other than this function, looks good. Please see my comment in #10412 - I believe that's wrong way to go about it. To some extend your finding confirms that different entities might have different needs RE protocol implementation (though I'm a bit surprised that Scribe is in this domain - I worry that it points to a potential bug somewhere).

File: server/routerlicious/api-report/protocol-base.api.md [hunk 2]
```diff
@@ -106,23 +130,42 @@ export function isSystemMessage(message: ISequencedDocumentMessage): boolean;
 export function mergeAppAndProtocolTree(appSummaryTree: ITree_2, protocolTree: ITree_2): ICreateTreeEntry[];
 
 // @public
-export class ProtocolOpHandler {
+export class ProtocolOpHandler implements IProtocolHandler {
```
Comment (vladsud) [inline server/routerlicious/api-report/protocol-base.api.md:133]: As a general rule, we should consider exporting only functor to create it, and not expose class itself. That's especially true with this work, though test code might want to look into guts.

File: server/routerlicious/api-report/protocol-base.api.md [hunk 3]
```diff
@@ -106,23 +130,42 @@ export function isSystemMessage(message: ISequencedDocumentMessage): boolean;
...
     // (undocumented)
+    setConnectionState(connected: boolean, clientId: string | undefined): void;
+    // (undocumented)
+    snapshot(): IQuorumSnapshot;
+    // (undocumented)
     readonly term: number;
 }
 
+// @public (undocumented)
+export class ProtocolOpHandlerWithClientValidation extends ProtocolOpHandler {
```
Comment (vladsud) [inline server/routerlicious/api-report/protocol-base.api.md:159]: Maybe this one should be called ProtocolOpHandler and above should be ProtocolOpHandlerEmptyQuorum?
Not sure

File: server/routerlicious/packages/protocol-base/src/protocol.ts [hunk 4]
```diff
@@ -136,7 +177,48 @@ export class ProtocolOpHandler {
...
             sequenceNumber: this.sequenceNumber,
             minimumSequenceNumber: this.minimumSequenceNumber,
-            ...this.quorum.snapshot(),
+            ...this._quorum.snapshot(),
         };
     }
 }
+
+export class ProtocolOpHandlerWithClientValidation extends ProtocolOpHandler {
+    constructor(
```
Comment (vladsud) [inline server/routerlicious/packages/protocol-base/src/protocol.ts:186]: I think you should be able to remove constructor and it should just work?
I'd remove "public" keyword from arguments below - they are already public in base class.

**Review by vladsud — APPROVED** [review]

File: server/routerlicious/api-report/protocol-base.api.md (hunk 2 above)
Comment (andre4i) [reply server/routerlicious/api-report/protocol-base.api.md:133]: Would something like this be viable? https://github.com/microsoft/FluidFramework/pull/10412/commits/ca0194dcd9bbeca1c405848f7f8b4fd8b10d46b7#diff-e4758e313a7e2cbeafa89bd3c08d5e48ffc2b2e205183e913faddaa021b5cf55R58

but we'd need to export both of them (at least for now, maybe in `next` we can remove it) as removing this class from exports is breaking the contract.

Comment (vladsud) [reply server/routerlicious/api-report/protocol-base.api.md:133]: Sure, though I'd drop ProtocolHandlerBuilder (type) - the shape of that function can be different for different implementations (unless we have to do type erasure and pass functor to Container, not actual implementation of interface).

Comment (andre4i) [reply server/routerlicious/api-report/protocol-base.api.md:133]: Was thinking this to be part of the loader. This package should give us the interface, then the loader has its own builder signature which people can implement.

**Pull request merged**